source/hpack.c

```diff
@@ -538,6 +524,105 @@ int aws_hpack_resize_dynamic_table(struct aws_hpack_context *context, size_t new
+    }
+
+    const size_t header_size = aws_hpack_get_header_size(header);
+
+    /* If for whatever reason this new header is bigger than the total table size, burn everything to the ground. */
+    if (AWS_UNLIKELY(header_size > context->dynamic_table.max_size)) {
+        return s_dynamic_table_shrink(context, 0);
+    }
+
+    /* If max size reached, rotate out headers until there's room for the new header */
+    if (s_dynamic_table_shrink(context, context->dynamic_table.max_size - header_size)) {
+        goto error;
+    }
+
+    /* If we're out of space in the buffer, grow it */
+    if (context->dynamic_table.num_elements == context->dynamic_table.buffer_capacity) {
+        /* If the buffer is currently of 0 size, reset it back to its initial size */
+        const size_t new_size =
+            context->dynamic_table.buffer_capacity
+                ? (size_t)(context->dynamic_table.buffer_capacity * s_hpack_dynamic_table_buffer_growth_rate)
+                : s_hpack_dynamic_table_initial_elements;
+
+        if (s_dynamic_table_resize_buffer(context, new_size)) {
+            goto error;
+        }
+    }
+
+    /* Decrement index 0, wrapping if necessary */
+    if (context->dynamic_table.index_0 == 0) {
+        context->dynamic_table.index_0 = context->dynamic_table.buffer_capacity - 1;
+    } else {
+        context->dynamic_table.index_0--;
+    }
+
+    /* Increment num_elements */
+    context->dynamic_table.num_elements++;
+    /* Increment the size */
+    context->dynamic_table.size += header_size;
+
+    /* Put the header at the "front" of the table */
+    struct aws_http_header *table_header = s_dynamic_table_get(context, 0);
+    *table_header = *header;
+
+    /* Write the new header to the look up tables */
+    if (aws_hash_table_put(
+            &context->dynamic_table.reverse_lookup, table_header, (void *)context->dynamic_table.index_0, NULL)) {
+        goto error;
+    }
+    /* Note that we can just blindly put here, we want to overwrite any older entry so it isn't accidentally removed. */
+    if (aws_hash_table_put(
+            &context->dynamic_table.reverse_lookup_name_only,
+            &table_header->name,
+            (void *)context->dynamic_table.index_0,
+            NULL)) {
+        goto error;
+    }
+
+    return AWS_OP_SUCCESS;
+
+error:
+    /* Do not attempt to handle the error, if something goes wrong, close the connection */
+    return AWS_OP_ERR;
+}
+
+int aws_hpack_resize_dynamic_table(struct aws_hpack_context *context, size_t new_max_size) {
+
+    /* Nothing to see here! */
+    if (new_max_size == context->dynamic_table.max_size) {
+        return AWS_OP_SUCCESS;
+    }
+
+    if (new_max_size > s_hpack_dynamic_table_max_size) {
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
+
+    /* If downsizing, remove elements until we're within the new size constraints */
+    if (s_dynamic_table_shrink(context, new_max_size)) {
+        goto error;
```

Comment:
maybe edit comment. sounds like you don't MEAN to call this unless we're downsizing, but we always call it because it's safe either way